nixos/mail/dovecot2: re-introduce extra settings and rename `sieveScripts`

RaitoBezarius opened the pull request:
## Description of changes

https://github.com/NixOS/nixpkgs/pull/275031 introduced structured configuration for the dovecot2 sieve plugin, by doing so, it broke SNM configuration doing Sieve configurations.

I also made `imapsieve.mailbox.*.causes` a list instead of an element.

I introduced a warning mechanism to give help to users who may have used `extraConfig` and will have it explode on them.

This attempts to fix up the public API to make it possible for SNM to pick up the pieces.

Here's patch for SNM, GitLab is too broken for me to open a PR over there:
```patch
From a15c4eef88d126ea9d02efbfbfdfd0fcbf9a9976 Mon Sep 17 00:00:00 2001
From: Raito Bezarius <masterancpp@gmail.com>
Date: Sun, 14 Jan 2024 20:55:20 +0100
Subject: [PATCH] dovecot: support new `sieve` API in nixpkgs

Since https://github.com/NixOS/nixpkgs/pull/275031 things have became more structured
when it comes to the sieve plugin.

Relies on https://github.com/NixOS/nixpkgs/pull/281001 for full
features.
---
 mail-server/dovecot.nix | 50 +++++++++++++++++++++--------------------
 1 file changed, 26 insertions(+), 24 deletions(-)

diff --git a/mail-server/dovecot.nix b/mail-server/dovecot.nix
index 7d73ee2..3e5f9fb 100644
--- a/mail-server/dovecot.nix
+++ b/mail-server/dovecot.nix
@@ -175,8 +175,14 @@ in
       mailPlugins.globally.enable = lib.optionals cfg.fullTextSearch.enable [ "fts" "fts_xapian" ];
       protocols = lib.optional cfg.enableManageSieve "sieve";
 
-      sieveScripts = {
-        after = builtins.toFile "spam.sieve" ''
+      pluginSettings = {
+        sieve = "file:${cfg.sieveDirectory}/%u/scripts;active=${cfg.sieveDirectory}/%u/active.sieve";
+        sieve_default = "file:${cfg.sieveDirectory}/%u/default.sieve";
+        sieve_default_name = "default";
+      };
+
+      sieve = {
+        scripts.after = builtins.toFile "spam.sieve" ''
           require "fileinto";
 
           if header :is "X-Spam" "Yes" {
@@ -184,8 +190,26 @@ in
               stop;
           }
         '';
+
+        pipeBins = [
+          pipeBin
+        ];
       };
 
+      imapsieve.mailbox = [
+        {
+          name = junkMailboxName;
+          causes = [ "COPY" "APPEND" ];
+          before = "${stateDir}/imap_sieve/report-spam.sieve";
+        }
+        {
+          name = "*";
+          from = junkMailboxName;
+          causes = [ "COPY" ];
+          before = "${stateDir}/imap_sieve/report-ham.sieve";
+        }
+      ];
+
       mailboxes = cfg.mailboxes;
 
       extraConfig = ''
@@ -307,28 +331,6 @@ in
           inbox = yes
         }
 
-        plugin {
-          sieve_plugins = sieve_imapsieve sieve_extprograms
-          sieve = file:${cfg.sieveDirectory}/%u/scripts;active=${cfg.sieveDirectory}/%u/active.sieve
-          sieve_default = file:${cfg.sieveDirectory}/%u/default.sieve
-          sieve_default_name = default
-
-          # From elsewhere to Spam folder
-          imapsieve_mailbox1_name = ${junkMailboxName}
-          imapsieve_mailbox1_causes = COPY,APPEND
-          imapsieve_mailbox1_before = file:${stateDir}/imap_sieve/report-spam.sieve
-
-          # From Spam folder to elsewhere
-          imapsieve_mailbox2_name = *
-          imapsieve_mailbox2_from = ${junkMailboxName}
-          imapsieve_mailbox2_causes = COPY
-          imapsieve_mailbox2_before = file:${stateDir}/imap_sieve/report-ham.sieve
-
-          sieve_pipe_bin_dir = ${pipeBin}/pipe/bin
-
-          sieve_global_extensions = +vnd.dovecot.pipe +vnd.dovecot.environment
-        }
-
         ${lib.optionalString cfg.fullTextSearch.enable ''
         plugin {
           plugin = fts fts_xapian
-- 
2.43.0
```

I also needed to add `services.dovecot2.sieve.extensions = [ "fileinto" ];`, I'm not sure why.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RaitoBezarius committed:
nixos/mail/dovecot2: re-introduce extra settings and rename `sieveScripts`

https://github.com/NixOS/nixpkgs/pull/275031 introduced structured configuration
for the dovecot2 sieve plugin, by doing so, it broke SNM configuration doing Sieve configurations.

This attempts to fix up the public API to make it possible for SNM to pick up the pieces.

RaitoBezarius commented:
cc @nlewo

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -473,6 +468,17 @@ in
+            sieve = "file:~/sieve;active=~/.dovecot.sieve";
+          }
+        '';
+        description = "Any extra settings to the dovecot2's sieve plugin, e.g. `sieve`, `sieve_default`, etc.";
```
```suggestion
        description = "Any extra plugin settings, e.g. `sieve`, `sieve_default`, etc.";
```

From what I can tell, the plugin block receives settings for any plugin, and I think it doesn't even matter if you put settings for different plugins into different plugin blocks or mix them. It would probably be better to move this out of sieve and make it "general purpose".

RaitoBezarius replied:
Hmm, sounds good.

RaitoBezarius replied:
I tried something, let me know, @2xsaiko, it's ideal to land this quickly otherwise we will break a lot of Dovecot2 setup out there :).

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -118,9 +110,11 @@ let
 
     ''
       plugin {
+        ${concatStringsSep "\n" (mapAttrsToList (to: from: "sieve_${to} = ${stateDir}/sieve/${to}") cfg.sieve.scripts)}
         sieve_plugins = ${concatStringsSep " " cfg.sieve.plugins}
         sieve_extensions = ${concatStringsSep " " (map (el: "+${el}") cfg.sieve.extensions)}
         sieve_global_extensions = ${concatStringsSep " " (map (el: "+${el}") cfg.sieve.globalExtensions)}
```
What do you think about making these and other plugin settings also use extraPluginSettings to reduce the amount of plaintext config?

RaitoBezarius replied:
This would remove many options, right? And make the migration significant for other users?

2xsaiko replied:
No, I'm talking about just setting `extraPluginSettings.sieve_plugins = "..."` in this module instead of putting it directly into the final config string. It shouldn't have significant (if any) impact on migration.

RaitoBezarius replied:
Hmm, why not yes, but in that case, I guess we ought to call it `pluginSettings`.

RaitoBezarius replied:
Kinda done.

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -473,6 +468,17 @@ in
       };
+
+      extraPluginSettings = mkOption {
+        type = types.attrsOf types.anything;
```
```suggestion
        type = types.attrsOf types.str;
```

Right now, this will only work for values that can be interpolated, so strings and attrsets with toString/outPath. Not sure if `str` covers those.

RaitoBezarius replied:
Done

RaitoBezarius committed:
nixos/mail/dovecot2: `imapsieve.mailbox.*.causes` is a list

Otherwise, it's not possible to pass `COPY,APPEND` properly.

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -405,14 +394,16 @@ in
             '';
-            example = "COPY";
-            type = types.nullOr (types.enum [ "APPEND" "COPY" "FLAG" ]);
+            example = literalExpression ''
+              [ "COPY" "APPEND" ]
+            '';
```
```suggestion
            example = [ "COPY" "APPEND" ];
```

RaitoBezarius replied:
Right!

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -116,33 +133,12 @@ let
-        cfg.sieve.pipeBins)}
-    '')
-    ''
+        ${concatStringsSep "\n" (mapAttrsToList (key: value: "${key} = ${value}") cfg.pluginSettings)}
```
```suggestion
      ${concatStringsSep "\n" (mapAttrsToList (key: value: "  ${key} = ${value}") cfg.pluginSettings)}
```

This should make indentation correct in the output file

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -376,6 +367,26 @@ in
+      # types.str does not coerce from packages, like `sievePipeBinScriptDirectory`.
+      type = types.attrsOf (types.oneOf [ types.str types.package ]);
+      default = {};
+      example = literalExpression ''
+        {
+          sieve = "file:~/sieve;active=~/.dovecot.sieve";
+        }
+      '';
```
```suggestion
      example = {
        sieve = "file:~/sieve;active=~/.dovecot.sieve";
      };
```

:)

RaitoBezarius replied:
I think this is kind of weird to have inconsistency on this in the file, see:
```
      example = literalExpression ''
        {
          Spam = { specialUse = "Junk"; auto = "create"; };
        }
      '';
```

I did it for the list, but it's weird to do it for the attribute set. What's the harm?

2xsaiko replied:
There isn't, it's just not necessary if it's just a simple expression that can't be evaluated any further afaik :P Keep it if you want.

RaitoBezarius replied:
I prefer to keep it this way and if we want to migrate this out, feel free to send follow up PRs? :p

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -501,6 +517,13 @@ in
       sieve.globalExtensions = optional (cfg.sieve.pipeBins != []) "vnd.dovecot.pipe";
+
+      pluginSettings = lib.mapAttrs (n: lib.mkDefault) ({
+        sieve_plugins = concatStringsSep " "cfg.sieve.plugins;
```
```suggestion
        sieve_plugins = concatStringsSep " " cfg.sieve.plugins;
```

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -4,14 +4,39 @@ let
+      name = "imapsieve_mailbox${toString idx}_after";
+      value = "file:${stateDir}/imapsieve/after/${baseNameOf el.after}";
+    }
+    ) cfg.imapsieve.mailbox));
```
```suggestion
  ) cfg.imapsieve.mailbox));
```

dblsaiko reviewed (changes requested):
Only nitpicks remain :P
Looks good, I haven't tested it yet though.

dotlambda commented:
> Here's patch for SNM, GitLab is too broken for me to open a PR over there:

Filed as https://gitlab.com/simple-nixos-mailserver/nixos-mailserver/-/merge_requests/316.

fpletz commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -501,6 +517,13 @@ in
+      pluginSettings = lib.mapAttrs (n: lib.mkDefault) ({
+        sieve_plugins = concatStringsSep " "cfg.sieve.plugins;
+        sieve_extensions = concatStringsSep " " (map (el: "+${el}") cfg.sieve.extensions);
+        sieve_global_extensions = concatStringsSep " " (map (el: "+${el}") cfg.sieve.globalExtensions);
```
By setting empty extensions and plugins, subsequent definitions of extensions and plugins done in `extraConfig` won't get picked up. You're basically breaking existing configs that were completely done in `extraConfig` before.

It broke my config and I had a hard time figuring out this was caused by the previous PR. This behavior change needs at least release notes but better not render these options at all if not defined explicitly.

RaitoBezarius replied:
You are totally right! :/

RaitoBezarius replied:
I think what we can do is to assert for the presence of those options inside the `extraConfig` string and err out if so to let people remediate to it or not.

RaitoBezarius replied:
What do you think about this strategy?

fpletz replied:
That's also fine by me. Though searching through `extraConfig` is a tad error-prone since there might be comments mentioning these options. But much better than weird dovecot2 errors about missing plugins when they are clearly defined. :smiley:

RaitoBezarius replied:
I tried something that takes into account comments, but I'm not really aware of the full grammar of Dovecot2 comments, so this is at most an OK solution.

RaitoBezarius committed:
nixos/mail/dovecot2: warn about potential collision due to structured configuration

Plugin configuration is pesky in dovecot2, let's warn about potential conflicts
in the module system by using a fancy regex.

This is only band-aid, this should be removed ASAP.

We clean up also a 21.05-era warning.

yu-re-ka commented:
IF everything can be configured via structured settings, why do we not remove the extraConfig (but with a proper mkRemovedOptionModule error)?
Then nixos-simple-mailserver can just adjust to use the structured settings.

2xsaiko commented:
> @2xsaiko In the future, it's my personal opinion, but I would prefer to avoid dealing with formatting nitpicks as those are pointless (in the presence of the formatting RFC which will trigger a treewide anyway) work that interrupts moving forward things, the other fix was… well.

Alright!

2xsaiko commented:
> IF everything can be configured via structured settings, why do we not remove the extraConfig (but with a proper mkRemovedOptionModule error)? Then nixos-simple-mailserver can just adjust to use the structured settings.

The structured settings right now, at least in this PR, are only for the plugin block, I'm going to have to look into the rest of the configuration syntax to see if it's possible to make it structured. From what I've seen it should be possible though.

yu-re-ka reviewed (approved):
Have not looked at the code in detail, but it's running on my mailserver (together with the SNM merge request) and seems to work fine

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -4,14 +4,67 @@ let
     mkEnableOption mkIf mkOption mkRemovedOptionModule optional optionalAttrs
-    optionalString singleton types;
+    optionalString singleton types mkRenamedOptionModule nameValuePair
+    mapAttrs' listToAttrs match filter;
```
lib.match doesn't seem to exist, not in master nor this commit.

```
% nix eval 'github:RaitoBezarius/nixpkgs/new-sieve-dovecot-api#lib.match'
error: flake 'github:RaitoBezarius/nixpkgs/new-sieve-dovecot-api' does not provide attribute 'packages.x86_64-linux.lib.match', 'legacyPackages.x86_64-linux.lib.match' or 'lib.match'
       Did you mean path?
```

RaitoBezarius replied:
Yeah, it's `lib.strings.match` unfortunately…

RaitoBezarius replied:
Fixed.

2xsaiko commented:
Nope, works fine here now.

chvp commented:
There still seem to be some problems, the generated dovecot conf contains these lines:
```
plugin {
    imapsieve_mailbox1_before = file:/var/lib/dovecot/imapsieve/before/report-spam.sieve
  imapsieve_mailbox1_causes = COPY,APPEND
  imapsieve_mailbox1_name = Junk
  imapsieve_mailbox2_before = file:/var/lib/dovecot/imapsieve/before/report-ham.sieve
  imapsieve_mailbox2_causes = COPY
  imapsieve_mailbox2_from = Junk
  imapsieve_mailbox2_name = *
  sieve = file:/var/sieve/%u/scripts;active=/var/sieve/%u/active.sieve
  sieve_after = /var/lib/dovecot/sieve//nix/store/gca1qh0ipsqf7c9h3z7jcagygcpkadbi-spam.sieve
  sieve_after2 = /var/lib/dovecot/sieve//nix/store/5126h7gj5jnmppj2ynxacw0yqibx2qd7-custom-spam.sieve
  sieve_default = file:/var/sieve/%u/default.sieve
  sieve_default_name = default
  sieve_extensions = +editheader +fileinto
  sieve_global_extensions = +vnd.dovecot.pipe
  sieve_pipe_bin_dir = /nix/store/bflj5daxrrfmlj64anfq5dydwd5xrfjb-sieve-pipe-bins
  sieve_plugins = sieve_imapsieve sieve_extprograms
}
```
The location for `sieve_after` and `sieve_after2` is not correct.

This is with this PR and [the SNM PR](https://gitlab.com/simple-nixos-mailserver/nixos-mailserver/-/merge_requests/316) applied.

2xsaiko commented:
@chvp could you try with #284749?

RaitoBezarius commented:
Good catch, the `from/to` natural variables there are confusing…
